### ISSUE TITLE
Question on clean reads before or after normalization

### ISSUE THREAD

**Issue by 000generic** [trinityrnaseq/trinityrnaseq#1218]
Hi!

I would like to use 1.6 billion Illumina reads for de novo transcriptome assembly by Trinity - and for genome scaffolding, with reads being mapped by HiSat2 to a genome pre-assembly for scaffolding.  

I decided to start with scaffolding  - and to use Trinity to normalize the reads prior to scaffolding on a machine with 64 CPU and 512 Gb RAM. I found that Trinity  throws errors and crashes when given too many reads - so I am bundling SRR data sets and normalizing individual bundles - typically 1-5 SRR data sets (there are over 400 total SRR data sets for the 1.6 billion reads).  A typical commandline line is:

`insilico_read_normalization.pl --seqType fq --JM 500G --max_cov 30 --left_list 1-pair_1-files --right_list 1-pair_2-files --pairs_together --PARALLEL_STATS --CPU 60`

I'm close to done with this - but have noticed a large percentage of reads are being discarded as aberrant. For example:

```
1774208 / 7889720 = 22.49% reads selected during normalization.
2463090 / 7889720 = 31.22% reads discarded as likely aberrant based on coverage profiles.
```

or 

```
815224 / 3140383 = 25.96% reads selected during normalization.
1819607 / 3140383 = 57.94% reads discarded as likely aberrant based on coverage profiles.
```

I have been running normalizations directly on downloaded SRR data - without cleaning (stand-alone Trimmomatic or in Trinity) or correcting (Rcorrector) the reads.  **I'm wondering (1)** if running things directly on uncleaned uncorrected reads was a mistake and is likely causing the high percentage of discarded reads. Or do the statistics above seem reasonable for stand-alone normalization by Trinity. 

Put another way, **I'm wondering (2)**, would you recommend Pipeline A or Pipeline B below for stand-alone large-scale normalization by Trinity:

Pipeline A
SRA download --> Trinity normalization

Pipeline B
SRA download --> Trimming --> Correcting --> Trinity normalization

**FInally I am wondering (3)**, given either pipeline, what settings or strategy would you recommend for dealing with so many reads / SRR data sets? I understand/believe I could have submitted all 400+ SRR datasets to Trinity, as Trinity can do single data set (SRR) normalizations and then combine them and normalize at end under the hood - and I probably should have done this rather than running separate jobs by hand - but I was also worried the combined one would crash things, as it would still be a large number of reads relative to what seems to complete error-free on my machine.

Thank you!
Eric

**Comment by brianjohnhaas** [comment]
Hi Eric,

Recent versions of Trinity's in silico normalization process would
discard reads as aberrant only if they had mean kmer coverage <= 0,
which I can envision happening only in the case where there's lots of
N characters or non-GATC scattered across the read sequences.
Earlier versions of Trinity's process would kick out aberrant reads as
those with high coefficient of variation in their kmer coverage
profiles.  That was effectively turned off because there were some
lowly expressed isoforms that we would retain if we didn't do that.

As a general approach, Id' go with normalizing the read sets
separately and then see if you can get that one last normalization in
that tackles all normalized reads. Even if you could do this with
handfuls of read sets at a time, it would be better than not doing it,
I think - at least easier on your total memory usage.

Wrt quality trimming, I'd suggest doing it pre-normalization.  That's
what Trinity does currently, if you have --trimmomatic set.

For massive data sets, using the --min_kmer_cov 2 or 3 would be a good
idea. (I wouldn't go higher than 3 here).

Let me know if I missed something here.  I'm definitely confused about
the high level of aberrants.

best,

~b

On Tue, Oct 25, 2022 at 3:43 PM Eric Edsinger ***@***.***> wrote:
>
> Hi!
>
> I would like to use 1.6 billion Illumina reads for de novo transcriptome assembly by Trinity - and for genome scaffolding, with reads being mapped by HiSat2 to a genome pre-assembly for scaffolding.
>
> I decided to start with scaffolding - and to use Trinity to normalize the reads prior to scaffolding on a machine with 64 CPU and 512 Gb RAM. I found that Trinity throws errors and crashes when given too many reads - so I am bundling SRR data sets and normalizing individual bundles - typically 1-5 SRR data sets (there are over 400 total SRR data sets for the 1.6 billion reads). A typical commandline line is:
>
> insilico_read_normalization.pl --seqType fq --JM 500G --max_cov 30 --left_list 1-pair_1-files --right_list 1-pair_2-files --pairs_together --PARALLEL_STATS --CPU 60
>
> I'm close to done with this - but have noticed a large percentage of reads are being discarded as aberrant. For example:
>
> 1774208 / 7889720 = 22.49% reads selected during normalization.
> 2463090 / 7889720 = 31.22% reads discarded as likely aberrant based on coverage profiles.
>
> or
>
> 815224 / 3140383 = 25.96% reads selected during normalization.
> 1819607 / 3140383 = 57.94% reads discarded as likely aberrant based on coverage profiles.
>
> I have been running normalizations directly on downloaded SRR data - without cleaning (stand-alone Trimmomatic or in Trinity) or correcting (Rcorrector) the reads. I'm wondering (1) if running things directly on uncleaned uncorrected reads was a mistake and is likely causing the high percentage of discarded reads. Or do the statistics above seem reasonable for stand-alone normalization by Trinity.
>
> Put another way, I'm wondering (2), would you recommend Pipeline A or Pipeline B below for stand-alone large-scale normalization by Trinity:
>
> Pipeline A
> SRA download --> Trinity normalization
>
> Pipeline B
> SRA download --> Trimming --> Correcting --> Trinity normalization
>
> FInally I am wondering (3), given either pipeline, what settings or strategy would you recommend for dealing with so many reads / SRR data sets? I understand/believe Trinity can do single data set (SRR) normalizations and then combine and normalize at end under the hood - and I probably should have done this rather than running separate jobs by hand - but I was also worried the combined one would crash things, as it would still be a large number of reads relative to what seems to complete error-free on my machine.
>
> Thank you!
> Eric
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>



-- 
--
Brian J. Haas
The Broad Institute
http://broadinstitute.org/~bhaas

**Comment by 000generic** [comment]
Hi b!

It turns out I was running Trinity 2.1.1 without realizing it.  I went ahead and finished the normalization process for all 1.6 billion reads using it - and then combined and normalized to produce a single set of around 55 million reads.  But given losses of 25-60%+ per subset due to aberrant discarded reads, I will probably do it again using the current 2.14 or 2.15 version of Trinity.  Can update here if that solves the issue.

**Comment by 000generic** [comment]
Good news!  Switching from version 2.1 to version 2.14 did the trick - for the same reads as above I now have zero losses due to aberrant reads:

2508699 / 3332865 = 75.27% reads selected during normalization.
0 / 3332865 = 0.00% reads discarded as likely aberrant based on coverage profiles.
0 / 3332865 = 0.00% reads discarded as below minimum coverage threshold=0

Bad news is I will have to rerun everything but that's not so bad - very happy to have things working!  I think I just forget to update things when I previously installed 2.14 - and had no idea there was already a much older 2.1 version on the machine. 

Thank you very much for your help - and sorry to have bothered you with something so trivial :)

**Comment by brianjohnhaas** [comment]
No worries.  I'd still be a bit worried about the high rate of reads
previously being identified as aberrant.  It could very well have to
do with quality trimming being needed.  Hopefully all goes smoothly,
though.  Let me know if it does give you any trouble.

best,

~b

On Thu, Oct 27, 2022 at 1:48 AM Eric Edsinger ***@***.***> wrote:
>
> Closed #1218 as completed.
>
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.Message ID: ***@***.***>



-- 
--
Brian J. Haas
The Broad Institute
http://broadinstitute.org/~bhaas
